### PR TITLE
[codex] Implement R10a harness evolution loop

### LINKED ISSUE #686: [Roadmap 10.8] Auto harness-evolution loop (R10a — model-agnostic, training-free)
## Context

Roadmap entry: **R10 was reframed (2026-04-29)** as two tracks — *R10a auto harness evolution* and *R10b fine-tuning*. R10.8 opens the R10a track. See the [roadmap diff](https://github.com/HybridAIOne/hybridclaw/blob/main/docs/content/internal/roadmap.md) and parent [R10 #553](https://github.com/HybridAIOne/hybridclaw/issues/553).

Anchored in **[Lin et al., *Agentic Harness Engineering*, arXiv:2604.25850v1, 2026-04-28](https://arxiv.org/html/2604.25850v1)**: 69.7% → 77.0% on Terminal-Bench 2 by editing skills / tools / middleware / memory from distilled traces — *no fine-tuning*. Frozen-harness transfer to weaker bases (DeepSeek / Qwen / Gemini) gave +5.1–10.1pp without re-evolution.

## Why a separate track from R10b SFT

| | R10a (this issue) | R10b (existing R10.4–R10.7) |
|---|---|---|
| Cost | LLM rollouts only | GPU training infra + data egress |
| Lead time | Hours per round | Days per fine-tune |
| Model lock-in | None — transfers across providers | Per-base-model |
| Storage | Git-tracked workspace + manifests | Tuned-model registry |
| Failure mode | Drift fixed by F12 rollback | Catastrophic forgetting |

Both consume the same **R10.1–R10.3** trajectory pipeline ✅.

## Loop (paper Algorithm 1)

\`\`\`
for round t in 1..T:
  1. rollout(k)        — k eval traces per task (paper used k=3–5)
  2. clean()           — strip base64, dedupe tool output
  3. attribute(t-1)    — verify previous F12 manifest predictions;
                          file-level rollback on disconfirmed edits
  4. distill()         — F11.4 Agent Debugger → 3-tier report
  5. evolve()          — evolve-agent edits surfaces under F12 contract
  6. commit()          — git tag; update H_best on pass@1 improvement
\`\`\`

## Acceptance criteria

- [ ] **R10.8.1** Evolve-agent harness (separate from the coworkers being improved): system prompt, tool to read F11.4 reports, tool to write to seven editable surfaces under F12 contract
- [ ] **R10.8.2** Workspace layout for the *target* coworker — one git directory per agent, seven mount points enforced by schema (system_prompt.md / tools.yaml / tools/ / middleware/ / sub_agents/ / config/ / long_term_memory/)
- [ ] **R10.8.3** Loop runner — orchestrates rollout → clean → attribute → distill → evolve → commit; configurable `T` (default 10), `k` (default 3), pass@1 + Succ/Mtok metrics ([paper Eq. 1 + 2](https://arxiv.org/html/2604.25850v1))
- [ ] **R10.8.4** Bash-only seed validator — refuses to start a fresh evolution from a non-minimal seed (paper §3.1 — prevents pre-fitted-config contamination); production coworkers can be evolved in-place but the metric must report seed delta
- [ ] **R10.8.5** Eval-suite plumbing — feeds A4 cost-regression gate; supports per-skill eval suites (R21 skills bring their own task set)
- [ ] **R10.8.6** Admin surface — round-by-round trajectory chart (pass@1, attribution score, edits per surface), drill into manifest entries (links to F12 admin surface)

## Hard prerequisites

- **F12 Change Manifest** [#685](https://github.com/HybridAIOne/hybridclaw/issues/685) — without falsifiable contracts the loop has no rollback signal
- **F11.4 Agent Debugger / subscriber pattern** [#623](https://github.com/HybridAIOne/hybridclaw/issues/623) — needs the 3-tier distilled output the evolve-agent reads
- **R10.1–R10.3** trajectory capture ✅ — already landed
- **R21.1** skill component contract — each evolvable skill must expose the seven surfaces

## Component-ablation guidance (paper §4.4.1)

Direct investment intuition for which surfaces the evolve-agent should touch first:

| Surface | Standalone Δpass@1 |
|---|---|
| Long-term memory | **+5.6pp** |
| Tools | +3.3pp |
| Middleware | +2.2pp |
| System prompt alone | **−2.3pp** (do not evolve in isolation) |

Components interact non-additively — co-evolve memory + tools + middleware before touching the prompt.

## Out of scope

- Cross-agent evolution (one loop = one coworker) — multi-agent coordination is a follow-up
- Real-time evolution during normal coworker operation — this is a *batch* loop run on the eval harness, not in production traffic
- Fine-tuning (R10b track)

## Open questions

1. Run cadence — nightly per skill? Per-customer per week?
2. Compute budget — paper used 10 rounds × ~hundreds of traces. Cap with A4.
3. Read-only enforcement on `runs/`, `verifier/`, model config — paper made this a hard invariant; we should too.

### PR DESCRIPTION
**Pull request by furukama**
Closes #686.

## Summary
- add the R10a harness-evolution core with seven editable surfaces, bash-only seed validation, command-backed rollouts, trace cleaning, F11.4-style report distillation, F12 manifest writes, attribution rollback, pass@1/Succ-Mtok metrics, cost gating, H_best tracking, and optional round commits
- add `hybridclaw harness-evolve init|validate-seed|run|status` for local batch operation
- expose authenticated admin run/manifest inspection through `/api/admin/harness-evolution` and add a console page at `/admin/harness-evolution`
- add unit coverage for workspace validation, F12 writes, rollback, metrics, dry loop summaries, and per-skill `evals/scenarios.json` command suites

## Validation
- `npx biome check --write src/evolution/harness-evolution.ts src/cli.ts src/cli/help.ts src/gateway/gateway-http-server.ts tests/harness-evolution.test.ts console/src/api/client.ts console/src/api/types.ts console/src/routes/harness-evolution.tsx console/src/router.tsx console/src/components/sidebar/navigation.ts`
- `./node_modules/.bin/vitest run --configLoader runner --project unit tests/harness-evolution.test.ts`
- `npm run lint`
- `npm --workspace console run test`

### CODE REVIEW

**Comment by furukama** [comment]
@claude Do a thorough review of this PR in terms of: KISS, YAGNI, DRY, Fail Fast, Secure by Default and performance/speed

**Comment by furukama** [comment]
Addressed the Claude review pass in the latest push (`cfc999a0`):

- KISS/DRY: extracted tools-array parsing, centralized harness flag parsing, removed duplicate manifest skeleton construction, read pre-edit file content once, and moved the shared metric display pattern to `Intl.NumberFormat` in both CLI rendering and the console.
- YAGNI/security: removed `metadata` from eval tasks and removed inline `beforeContent` from F12 manifests; rollback snapshots now live in per-round sidecars instead of the API-visible manifest.
- Fail-fast: validates fixed `reportPath` before the loop and stops additional rounds once the suite cost budget is exceeded.
- Secure by default: replaced `shell: true` command execution with argv-based `spawn`, hardened gateway summary/manifest reads with realpath checks and cached allowlisted roots, and replaced `git add .` with explicit harness paths.
- Performance: writes F12 manifests once per round, writes/reads lightweight `list-entry.json`, runs command-backed rollouts concurrently, and caps trace text before splitting/deduping.

Validation:
- `./node_modules/.bin/vitest run --configLoader runner --project unit tests/harness-evolution.test.ts`
- `npm run lint`

Claude's review was a top-level PR comment, so there was no Claude review thread to resolve. Existing review threads are resolved.

**Comment by furukama** [comment]
Addressed the remaining AC gaps in `794a09c6`:

- Wired the closed-loop evolve step: when a round has no injected edits and the distilled report has no fenced F12 JSON, the loop now calls `runEvolveAgent()`, which uses the auxiliary model layer with `EVOLVE_AGENT_SYSTEM_PROMPT` and the F12 tool schema, writes `evolve-agent-output.md`, parses returned JSON edits, and applies them through the F12 surface writer.
- Added deterministic test injection for the evolve-agent runner and coverage proving a no-JSON distilled report now produces/applies edits through the evolve path.
- Added `seedDelta` to `EvolutionRunResult` and saved summaries for in-place production coworker runs, including changed surface count, changed surfaces, file count, and notes.
- Exposed seed delta and evolve-source metadata in console/API types and the harness evolution admin view.

Validation:
- `./node_modules/.bin/vitest run --configLoader runner --project unit tests/harness-evolution.test.ts` (15 passed)
- `npm run lint`
- `npm --workspace console run test` (349 passed; existing jsdom navigation warning, exit 0)

**Comment by furukama** [comment]
Fixed the remaining nits in `a90ee698`:

- Aligned the loop order with the paper: rollout → clean → attribute → distill → evolve → commit.
- Added programmatic component ordering before F12 writes so long-term memory/tool/middleware edits are applied before system prompt edits.
- Added a visible pass@1 trajectory chart above the rounds table in the admin view.
- Added regression coverage for edit ordering.

Validation:
- `./node_modules/.bin/vitest run --configLoader runner --project unit tests/harness-evolution.test.ts` (16 passed)
- `npm run lint`
- `npm --workspace console run test` (349 passed; existing jsdom navigation warning, exit 0)
